dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java

```diff
@@ -297,5 +652,1699 @@ void setContextualProperty(String key, Object v) {
+                ret = getAsyncPostDispatch();
+                setAsyncPostDispatch(value);
+            } else if (key == SECURITY_CONTEXT) {
+                ret = getSecurityContext();
+                setSecurityContext(value);
+            } else if (key == AUTHORIZATION_POLICY) {
+                ret = getAuthorizationPolicy();
+                setAuthorizationPolicy(value);
+            } else if (key == CERT_CONSTRAINTS) {
+                ret = getCertConstraints();
+                setCertConstraints(value);
+            } else if (key == AbstractHTTPDestination.SERVICE_REDIRECTION) {
+                ret = getServiceRedirection();
+                setServiceRedirection(value);
+            } else if (key == HTTP_SERVLET_RESPONSE) {
+                ret = getHttpServletResponse();
+                setHttpServletResponse(value);
+            } else if (key == RESOURCE_METHOD) {
+                ret = getResourceMethod();
+                setResourceMethod(value);
+            } else if (key == ONE_WAY_REQUEST) {
+                ret = getOneWayRequest();
+                setOneWayRequest(value);
+            } else if (key == ASYNC_RESPONSE) {
+                ret = getAsyncResponse();
+                setAsyncResponse(value);
+            } else if (key == THREAD_CONTEXT_SWITCHED) {
+                ret = getThreadContextSwitched();
+                setThreadContextSwitched(value);
+            } else if (key == OutgoingChainInterceptor.CACHE_INPUT_PROPERTY) {
+                ret = getCacheInputProperty();
+                setCacheInputProperty(value);
+            } else if (key == PhaseInterceptorChain.PREVIOUS_MESSAGE) {
+                ret = getPreviousMessage();
+                setPreviousMessage(value);
+            } else if (key == AbstractHTTPDestination.RESPONSE_HEADERS_COPIED) {
+                ret = getResponseHeadersCopied();
+                setResponseHeadersCopied(value);
+            } else if (key == SSE_EVENT_SINK) {
+                ret = getSseEventSink();
+                setSseEventSink(value);
+            } else if (key == REQUESTOR_ROLE) {
+                ret = getRequestorRole();
+                setRequestorRole(value);
+            } else if (key == PARTIAL_RESPONSE_MESSAGE) {
+                ret = getPartialResponse();
+                setPartialResponse(value);
+            } else if (key == EMPTY_PARTIAL_RESPONSE_MESSAGE) {
+                ret = getEmptyPartialResponse();
+                setEmptyPartialResponse(value);
+            } else if (key == ENDPOINT_ADDRESS) {
+                ret = getEndpointAddress();
+                setEndpointAddress(value);
+            } else if (key == INBOUND_MESSAGE) {
+                ret = getInboundMessage();
+                setInboundMessage(value);
+            }
+            return ret == NOT_FOUND ? null : ret;
+        }
+
+        return super.put(key, value);
+    }
+
+    @Override
+    public Set<String> keySet() {
```

Comment:
This applies to keySet, entrySet and values.  Do we want to support this calling Iterator.remove methods on the returned Collections?  If so it would be a custom implementation of the Collection returned and the Iterator would be able to track where we are with the current key, value, entry and be able to do the remove method based off of that.  With us having an array of getter/setter properties it will be easier to track what should be removed on the Collection iterator.

If we keep this current mechanism, then we should make a copy of the Set that is returned for super.keySet() like you do later one so we aren't modifying the parents returned Set object.

Reply:
So you decided to keep the current mechanism and not support Iterator.remove on the Set / Collection returned from keySet, entrySet and values?  If we do that we probably should have the Iterator remove return UnsupportedOperationException.  This is a change in behavior from the previous implementation of this classes where the Iterator would actually work.  I don't know if anyone calls keySet, entrySet or values on the MessageImpl class.